src/dune_lang/decoder.ml

```diff
@@ -384,6 +386,47 @@ let triple a b c =
       b >>= fun b ->
       c >>= fun c -> return (a, b, c) )
 
+let unit_number name suffixes =
+  let unit_number_of_string ~loc s =
+    let l = String.length s in
+    let n, suffix =
+      let rec suffix_length i =
+        if i = l then
+          i
+        else
+          let c = s.[l - i - 1] in
+          if Char.code c >= Char.code '0' && Char.code c <= Char.code '9' then
+            i
+          else
+            suffix_length (i + 1)
```

Comment:
How about replacing this recursive function by implementing this one (taken from base):

```ocaml
(** [lfindi ?pos t ~f] returns the smallest [i >= pos] such that [f i t.[i]], if there is
    such an [i].  By default, [pos = 0]. *)
val lfindi : ?pos:int -> t -> f:(int -> char -> bool) -> int option
```

Another option is:

```ocaml
val lsplit2_on_match : string -> f:(char -> bool) -> (string * string) option
```

Reply:
I took a look at lfindi and it's (tail) recursive too: ```
let lfindi ?(pos=0) t ~f =
  let n = length t in
  let rec loop i =
    if i = n then None
    else if f i t.[i] then Some i
    else loop (i + 1)
  in
  loop pos
```

Unless you meant you just wanted me to give it the same interface ?

Reply:
Feel free to pick whatever interface/implementation you'd like here. I just don't like seeing low-level code in the application logic. It makes it hard for me to understand the intention of the code.